Partial adoption of dynamicDowncast and RefPtr in AccessibilityNodeObject

annevk opened the pull request:
#### 8797498dccaef3a0f69193b93c8235eb65ae1381
<pre>
Partial adoption of dynamicDowncast and RefPtr in AccessibilityNodeObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=266170">https://bugs.webkit.org/show_bug.cgi?id=266170</a>

Reviewed by Chris Dumez.

* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::canvasHasFallbackContent const):
(WebCore::AccessibilityNodeObject::isNativeTextControl const):
(WebCore::AccessibilityNodeObject::isSearchField const):
(WebCore::AccessibilityNodeObject::isNativeImage const):
(WebCore::AccessibilityNodeObject::isInputImage const):
(WebCore::AccessibilityNodeObject::isEnabled const):
(WebCore::AccessibilityNodeObject::isPressed const):
(WebCore::AccessibilityNodeObject::isChecked const):
(WebCore::AccessibilityNodeObject::isHovered const):
(WebCore::AccessibilityNodeObject::isMultiSelectable const):
(WebCore::AccessibilityNodeObject::isRequired const):
(WebCore::AccessibilityNodeObject::valueForRange const):
(WebCore::AccessibilityNodeObject::maxValueForRange const):
(WebCore::AccessibilityNodeObject::minValueForRange const):
(WebCore::isNodeActionElement):
(WebCore::AccessibilityNodeObject::labelElementContainer const):
(WebCore::AccessibilityNodeObject::alternativeText const):
(WebCore::AccessibilityNodeObject::visibleText const):
(WebCore::AccessibilityNodeObject::url const):
(WebCore::AccessibilityNodeObject::title const):
(WebCore::AccessibilityNodeObject::text const):
(WebCore::AccessibilityNodeObject::stringValue const):
(WebCore::AccessibilityNodeObject::colorValue const):
(WebCore::AccessibilityNodeObject::canSetFocusAttribute const):
(WebCore::AccessibilityNodeObject::canSetValueAttribute const):

Canonical link: <a href="https://commits.webkit.org/271858@main">https://commits.webkit.org/271858@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7adc003e473f783d3bfb01993f039df1a5d14d02

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29843 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31156 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32353 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26984 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30499 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10703 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5770 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27038 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30133 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7138 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25453 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6075 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6259 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26539 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33689 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27234 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26964 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32418 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6162 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4357 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30201 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7906 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7078 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6914 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6693 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -352,11 +352,9 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
-        return selectElement.multiple() ? AccessibilityRole::ListBox : AccessibilityRole::PopUpButton;
-    }
-    if (is<HTMLImageElement>(*node()) && downcast<HTMLImageElement>(*node()).hasAttributeWithoutSynchronization(usemapAttr))
+    if (RefPtr selectElement = dynamicDowncast<HTMLSelectElement>(node()))
```
`*node()` to avoid an unnecessary null check.

cdumez commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -352,11 +352,9 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
-    if (is<HTMLImageElement>(*node()) && downcast<HTMLImageElement>(*node()).hasAttributeWithoutSynchronization(usemapAttr))
+    if (RefPtr selectElement = dynamicDowncast<HTMLSelectElement>(node()))
+        return selectElement->multiple() ? AccessibilityRole::ListBox : AccessibilityRole::PopUpButton;
+    if (RefPtr imgElement = dynamicDowncast<HTMLImageElement>(node()); imgElement && imgElement->hasAttributeWithoutSynchronization(usemapAttr))
```
`*node()`

cdumez commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -368,30 +366,29 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
-    if (is<HTMLInputElement>(node())) {
-        HTMLInputElement& input = downcast<HTMLInputElement>(*node());
-        if (input.isSwitch())
+    if (RefPtr input = dynamicDowncast<HTMLInputElement>(node())) {
```
`*node()`

cdumez commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -531,7 +528,7 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
     if (supportsARIAAttributes() || canSetFocusAttribute())
         return AccessibilityRole::Group;
-    if (is<Element>(*node()) && downcast<Element>(*node()).isFocusable())
+    if (RefPtr element = dynamicDowncast<Element>(node()); element && element->isFocusable())
```
`*node()`.

cdumez commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -688,26 +685,19 @@ bool AccessibilityNodeObject::computeAccessibilityIsIgnored() const
-    if (!is<HTMLCanvasElement>(node))
-        return false;
-    HTMLCanvasElement& canvasElement = downcast<HTMLCanvasElement>(*node);
+    RefPtr canvasElement = dynamicDowncast<HTMLCanvasElement>(node());
```
The previous code was null-checking the node so passing `node()` here is correct.

cdumez commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1254,9 +1226,8 @@ bool AccessibilityNodeObject::toggleDetailsAncestor()
-    if (is<HTMLInputElement>(*node)) {
-        HTMLInputElement& input = downcast<HTMLInputElement>(*node);
-        if (!input.isDisabledFormControl() && (input.isRadioButton() || input.isCheckbox() || input.isTextButton() || input.isFileUpload() || input.isImageButton()))
+    if (RefPtr input = dynamicDowncast<HTMLInputElement>(node)) {
```
`*node` to match the previous code.

cdumez commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1899,8 +1870,8 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
     // The fieldset element derives its alternative text from the first associated legend element if one is available.
-    if (objectCache && is<HTMLFieldSetElement>(*node)) {
-        AccessibilityObject* object = objectCache->getOrCreate(downcast<HTMLFieldSetElement>(*node).legend());
+    if (RefPtr fieldset = dynamicDowncast<HTMLFieldSetElement>(node); fieldset && objectCache) {
```
`*node`

cdumez reviewed (approved):
LGTM with nits to avoid new null checks.

annevk committed:
Partial adoption of dynamicDowncast and RefPtr in AccessibilityNodeObject
https://bugs.webkit.org/show_bug.cgi?id=266170

Reviewed by Chris Dumez.

* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::canvasHasFallbackContent const):
(WebCore::AccessibilityNodeObject::isNativeTextControl const):
(WebCore::AccessibilityNodeObject::isSearchField const):
(WebCore::AccessibilityNodeObject::isNativeImage const):
(WebCore::AccessibilityNodeObject::isInputImage const):
(WebCore::AccessibilityNodeObject::isEnabled const):
(WebCore::AccessibilityNodeObject::isPressed const):
(WebCore::AccessibilityNodeObject::isChecked const):
(WebCore::AccessibilityNodeObject::isHovered const):
(WebCore::AccessibilityNodeObject::isMultiSelectable const):
(WebCore::AccessibilityNodeObject::isRequired const):
(WebCore::AccessibilityNodeObject::valueForRange const):
(WebCore::AccessibilityNodeObject::maxValueForRange const):
(WebCore::AccessibilityNodeObject::minValueForRange const):
(WebCore::isNodeActionElement):
(WebCore::AccessibilityNodeObject::labelElementContainer const):
(WebCore::AccessibilityNodeObject::alternativeText const):
(WebCore::AccessibilityNodeObject::visibleText const):
(WebCore::AccessibilityNodeObject::url const):
(WebCore::AccessibilityNodeObject::title const):
(WebCore::AccessibilityNodeObject::text const):
(WebCore::AccessibilityNodeObject::stringValue const):
(WebCore::AccessibilityNodeObject::colorValue const):
(WebCore::AccessibilityNodeObject::canSetFocusAttribute const):
(WebCore::AccessibilityNodeObject::canSetValueAttribute const):

Canonical link: https://commits.webkit.org/271858@main
